Disable apply button when there is no change to selected filters

imanish003 opened the pull request:
This PR fixed issue #7116, in which Apply button behavior of **Filter by Attribute** block is not as expected. 

### What is the issue?
- If the user selects a few attributes & saves the result using Apply button, then Apply button remains active. The expected result is that Apply button should become disabled after saving the result.
- After saving the result, even if we refresh the page, Apply button remains active. The expected result is that Apply button should remain disabled unless the user makes some change to active filters.

Please find a little more info in issue #7116 

### What does this change do?
In this PR, I implemented a new function, `getIsApplyButtonDisabled`, which returns if Apply button should be disabled or not.


Fixes #7116 

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Create a new post with an `All products` & `Filter by Attribute` block. Please make sure that Apply filter button is enabled, as shown in screenshot below:
![image](https://user-images.githubusercontent.com/16707866/195314106-00a753e1-7c37-4a31-9e8a-bd63d44bff21.png) 

2. After saving the post, Frontend should show Apply button as disabled as initially no filter is selected, as shown in screenshot below:
![image](https://user-images.githubusercontent.com/16707866/195316952-c6e54499-02dd-41bd-8e13-1e478ba7f00e.png)


3. As soon as you check any option, Apply button should become active & clickable.
![image](https://user-images.githubusercontent.com/16707866/195316991-daf497ca-0510-4018-b006-53c30f44be50.png)

4. After saving the result using the Apply button, Apply button should become disabled:
![image](https://user-images.githubusercontent.com/16707866/195317052-bfb55864-1116-43db-aef8-cb99af1faeb6.png)

5. Now refresh the page. Even after refreshing the page, Apply button should remain disabled until you change the selected options.

6. Perform all the above testing steps for Dropdown display style too as shown in screenshot below
![image](https://user-images.githubusercontent.com/16707866/195317095-7c689f15-8a7f-4527-be81-aea836fb929e.png)
![image](https://user-images.githubusercontent.com/16707866/195317119-f93fa3c3-a29e-483f-ac40-38de48fab0bb.png)




* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [X] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

imanish003 committed:
Disable apply button when there is no change to selected options

More info: https://github.com/woocommerce/woocommerce-blocks/issues/7116

imanish003 committed:
Merge branch 'trunk' into fix/7116-disable-apply-button

thealexandrelara committed:
Add tests to the Filter Attribute block

Since we are making changes to the Filter Attributes block to fix issue #7373, we are also adding tests to make sure the changes are providing the expected behavior

The setup-globals file was changed to add 'attributes' and 'has_filterable_products' properties to it, so it can be used in the tests.

thealexandrelara committed:
Merge test utils methods with attribute filter block

Removing test utils file and merging the methods with the Attribute Filter block to prevent tests from failing due to Jest configuration that only allow test files to exists inside a 'test' folder

tjcafferkey commented on assets/js/blocks/attribute-filter/block.tsx:
```diff
@@ -536,6 +536,21 @@ const AttributeFilterBlock = ( {
 	setWrapperVisibility( true );
 
+	const getIsApplyButtonDisabled = () => {
+		if ( termsLoading || countsLoading || checked.length === 0 ) {
```
What about if I have one attribute checked, and I _uncheck_ this and go to click "Apply" assuming this would then have the same effect as the "Clear all" functionality.

Presumably based on this line, the "Apply" button will be disabled because `checked.length` will be `0`?

imanish003 replied:
@tjcafferkey thanks for pointing that out 🙏  As this condition was already there, I didn't touch it. But Looks like we don't need `checked.length === 0` condition anymore. 

**Why was this condition needed before?**
Because when the user initially load the page, no filter is selected therefore we needed this condition to keep the Apply disabled in this case.

**Why we don't need it anymore?**
Because now in this PR we add following code:
```jsx
const activeFilters = getActiveFilters( attributeObject );
if ( activeFilters.length === checked.length ) {
	return checked.every( ( value ) =>
		activeFilters.includes( value )
	);
}
```
This will return `true` when the user initially loads the page & Apply button will become disabled. Therefore, I believe it's safe to remove `checked.length === 0` condition.

I will remove this condition 👍

imanish003 replied:
@tjcafferkey FYI pushed the changes.

imanish003 committed:
Remove checked.length === 0 condition

**Why was this condition needed before?**
Because when the user initially load the page, no filter is selected therefore we needed this condition to keep the Apply disabled in this case.

**Why we don't need it anymore?**
Because now in this PR we add following code:
```jsx
const activeFilters = getActiveFilters( attributeObject );
if ( activeFilters.length === checked.length ) {
	return checked.every( ( value ) =>
		activeFilters.includes( value )
	);
}
```
This will return `true` when the user initially loads the page & Apply button will become disabled. Therefore, I believe it's safe to remove `checked.length === 0` condition.

More info: https://github.com/woocommerce/woocommerce-blocks/pull/7373#discussion_r994737313

imanish003 commented on tests/js/setup-globals.js:
```diff
@@ -15,6 +15,7 @@ global.wcSettings = {
 	date: {
 		dow: 0,
 	},
+	has_filterable_products: true,
```
@thealexandrelara Thanks for writing the tests 🙌🏻 I am still getting familiar with the way we write tests, but I think it should be camelCase instead of snake_case to be consistent with other field names in this file.

@tjcafferkey What do you think? 🙂

tjcafferkey replied:
@imanish003 I believe @thealexandrelara is mocking [this setting value](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/blocks/attribute-filter/block.tsx#L86-L90) which is what the block uses internally to decide whether to render or not. [This value gets set server-side](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTypes/AllProducts.php#L24-L25) (in other block files too) and passed to the client where it is accessed on the global `wcSettings` object, to tell the filters there are products on the current page which are filterable.

This was done because filter blocks can be used in Classic themes (non-block themes) such as Storefront in widget areas which means they would get rendered on pages where no products would be present (such as a T&C's page for example). So we set this value so we can return `null` in the filter when there are no products available to render.

tjcafferkey replied:
You can read more about this specific functionality on this PR: https://github.com/woocommerce/woocommerce-blocks/pull/6270

imanish003 replied:
@tjcafferkey Ah! Got it. Thanks for explaining that 🙏🏻

tjcafferkey reviewed (approved):
This works great, good work! LGTM

imanish003 committed:
Merge branch 'trunk' into fix/7116-disable-apply-button

imanish003 commented:
Hi @thealexandrelara, It Looks like one [test is failing](https://github.com/woocommerce/woocommerce-blocks/actions/runs/3248768780/jobs/5330413630) after making the changes [discussed here](https://github.com/woocommerce/woocommerce-blocks/pull/7373#discussion_r994737313). Can you please update the test? After that, I think we can merge the branch with the trunk 🙂

thealexandrelara commented:
> Hi @thealexandrelara, It Looks like one [test is failing](https://github.com/woocommerce/woocommerce-blocks/actions/runs/3248768780/jobs/5330413630) after making the changes [discussed here](https://github.com/woocommerce/woocommerce-blocks/pull/7373#discussion_r994737313). Can you please update the test? After that, I think we can merge the branch with the trunk 🙂

hey @imanish003, I think the test is failing because now if a user enters the page with a filter attribute selected, for example, Large (for the Filter by Size attribute block) and then they remove that selected attribute, the Apply button will continue to be disabled and they won't be allowed to remove all the filters. Here is a gif showing that:

![Screen Recording 2022-10-14 at 08 57 16](https://user-images.githubusercontent.com/20469356/195847624-c76a209a-7a86-4aa9-89db-8e81d6adb889.gif)

One suggestion of what we could do is to move the `checked` and `activeFilters` part to a method, for instance:
```
 const checkIfAppliedFiltersAreEqualCheckedFilters = () => {
        const currentAppliedFilters = getActiveFilters( attributeObject );
        if ( checked.length !== currentAppliedFilters.length ) return false;

        return checked.every( ( checkedFilter ) => {
            return currentAppliedFilters.includes( checkedFilter );
        } );
  };
```

Then we can store the condition for enabling/disabling the apply button in a variable. In this case, we don't remove the `checked.length === 0` condition because it is checking if :
```
const disableApplyButton = termsLoading ||
 							countsLoading ||
 							checked.length === 0 ||
                                                        checkIfAppliedFiltersAreEqualCheckedFilters()
```
And finally we can use that variable to disable the button:
```
<FilterSubmitButton
						className="wc-block-attribute-filter__button"
						isLoading={ isLoading }
						disabled={ disableApplyButton }
						onClick={ () => onSubmit( checked ) }
					/>
```

If you prefer not to split in a separate function, you could just invert the logic a little bit inside the `getIsApplyButtonDisabled` method. This way we can keep the previous conditions but also add the check for when the applied filters are changed:
```
const getIsApplyButtonDisabled = () => {
		const activeFilters = getActiveFilters( attributeObject );
		let hasAppliedFiltersChanged = false;
		if ( activeFilters.length === checked.length ) {
			hasAppliedFiltersChanged = checked.every( ( value ) =>
				activeFilters.includes( value )
			);
		}

		return (
			termsLoading ||
			countsLoading ||
			checked.length === 0 ||
			hasAppliedFiltersChanged
		);
	};
```
I hope it helps 🙌

imanish003 commented:
Hi @thealexandrelara,
This seems to be working fine for me, as you can see in the video below:
https://user-images.githubusercontent.com/16707866/196139750-b9f2e4af-8fd9-479c-aaf1-110c7fdee61c.mov

Did you try it on the branch `fix/7116-disable-apply-button`? 🙂

thealexandrelara committed:
Add filterSize param to test setup

With this configuration, it is now possible to modify the current filter size in order to improve tests

imanish003 commented:
@thealexandrelara Looks good now. Merging with `trunk` 🙌🏻

imanish003 committed:
Merge branch 'trunk' into fix/7116-disable-apply-button
